### DISCUSSION TITLE
Jotai problems with micro front ends

### DISCUSSION THREAD

**Discussion by Aubrey-Russell** [pmndrs/jotai#2521] [Bug report]
If you have multiple micro front ends that all use Jotai for state management, it causes all sort of conflicts and race conditions due to (my theory) conflicting react contexts, particularly with atomWithStorage. Even wrapping all MFEs with a provider does not help. I've found that using:

```
const store = createStore();

<Provider store={store}>
```

does solve the problem, but this was not obvious at all how to solve the issue and it seems like Jotai's reliance on react contexts might be causing issues with MFE architectures. I don't really understand the nuance of what happens when multiple react apps with federated modules are loaded into web pack that depend on the same jotai package. When jotai provider goes to use getDefaultStore or createContext, what happens? 

I do see a previous discussion with @Cpt-Falcon, but it was more of a mitigation compared to a real solution

Online my research indicated this:

"In Webpack federated modules, React contexts can conflict when multiple copies of a library exist and one context instance is created while another is accessed"

**Comment by dai-shi** [comment]
I'm not familiar with MFEs, but wouldn't it be something that causes without MFEs?
Because if `<Provider store={store}>` works, it sounds like default store works too. Or, if it's a MFE thing, `<Provider store={store}>` seems like a right solution.

You might be interested in reading the Provider code. It's short.

https://github.com/pmndrs/jotai/blob/7a68a1246f1c174af3dd2979de202aa6869db7a7/src/react/Provider.ts#L21-L39
